### DISCUSSION TITLE
Proposal: Repeating action

### DISCUSSION THREAD

**Discussion by mosesnagar** [dotnet/runtime#63223] [Ideas]
Hi all,

posted it on csharplang, and it seems it related to this repo.

sometime you need repeat some code for X times. I mean code like that:
`
for(var i=0; i<X;i++) {
  ....
  var x = SomeGenerator();
  SomeAction(x);
  ......
}
`

I found some ways to do it online, for example:
`Enumerable.Repeat<Action>(() => { Console.WriteLine("Hello World"); this.DoSomething(); },10).ToList().ForEach(f => f.Invoke());`
`
Enumerable.Repeat<Action>(() => 
{
    Console.WriteLine("Hello World.");
    this.DoSomeStuff();
}, 10).ToList().ForEach(x => x());
`
and more as you can see here for example: 
https://stackoverflow.com/questions/6406741/is-there-an-elegant-way-to-repeat-an-action

but I don't find it very elegant. I thought about method extension for action, so we can do:
`SomeAction().Repeat(times)` or `Repeat(aciton, times) or Repeat(() => { ... }, times)` or `3.Repeat(()=>{ ... });`

WDYT?

**Comment by huoyaoyuan** [comment]
LINQ was first designed for side-effect free operations. That's why the don't add `ForEach` methods.

**Reply by mosesnagar** [reply]
first, thanks for your input!
as far as I understand, LINQ it's operations on data. I'm suggesting a static method like `Repeat(() => { DoSomething(); }, 2)` or  `Retry(() => { DoSomething(); }, 2)`. Or an extension method on Func\Action and not on data.

**Reply by Clockwork-Muse** [reply]
> first, thanks for your input! as far as I understand, LINQ it's operations on data. I'm suggesting a static method like `Repeat(() => { DoSomething(); }, 2)` or `Retry(() => { DoSomething(); }, 2)`. Or an extension method on Func\Action and not on data.

This is implying you're calling an action for a side-effect, which is usually not a good idea.  It would be far better to arrange things so that whatever repeated code you have works off of input data, thus allowing it to work with LINQ.

Can you provide a (possibly anonymized) real-world use-case?
